### PR TITLE
Minimize re-renders during voice conversation

### PR DESCRIPTION
**Pull request by crui3er**
- changed how author/account presence rendered;
- minimize re-renders for unread chat counters;
- miniize re-renders in AuthorList;

### CODE REVIEW

**Review by alexis-kochetov — APPROVED** [review]

File: src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListTabUnreadCount.razor [hunk 1]
```diff
@@ -42,10 +44,12 @@
         _ = History.NavigateTo(Links.Chat(firstUnreadChat.Id));
     }
 
+    protected override bool ShouldRender()
+        => State.HasError || State.Value != _rendered;
+
     public sealed record Model {
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/ChatList/ChatListTabUnreadCount.razor:50]: Can it be record struct?

File: src/dotnet/Chat.UI.Blazor/Components/ChatRightPanel/AuthorList.razor [hunk 2]
```diff
@@ -41,14 +41,25 @@
...
+    private EventCallback<AuthorId> _onClickCallback;
+
     private Session Session => ChatContext.Session;
     private Chat Chat => ChatContext.Chat;
     private IRoles Roles => ChatContext.Roles;
     private IAuthors Authors => ChatContext.Authors;
     private AuthorUI AuthorUI => ChatContext.AuthorUI;
 
+    [Inject] private ILogger<AuthorList> Log { get; init; } = null!;
     [Parameter, EditorRequired] public ChatContext ChatContext { get; set; } = null!;
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/ChatRightPanel/AuthorList.razor:53]: Missing empty line.

File: src/dotnet/Chat.UI.Blazor/Components/Author/AuthorPresenceIndicator.razor [hunk 3]
```diff
@@ -0,0 +1,70 @@
+﻿@namespace ActualChat.Chat.UI.Blazor.Components
+@inherits ComputedStateComponent<Presence>
+
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/Author/AuthorPresenceIndicator.razor:3]: No empty line is needed here.

File: src/dotnet/Chat.UI.Blazor/Components/Navbar/ChatsNavbarButtonBadge.razor [hunk 4]
```diff
@@ -34,5 +26,8 @@
         return new Model(count);
     }
 
+    protected override bool ShouldRender()
+        => State.HasError || State.Value != _rendered;
+
     public sealed record Model(Trimmed<int> Count = default);
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/Navbar/ChatsNavbarButtonBadge.razor:32]: Record struct?

File: src/dotnet/UI.Blazor/Components/Account/AccountBadgeBase.cs [hunk 5]
```diff
@@ -30,17 +28,10 @@ protected override ComputedState<Model>.Options GetStateOptions()
...
-
-        return new(account, presence);
+        return new(account);
     }
 
-    public sealed record Model(
-        Account Account,
-        Presence Presence = Presence.Unknown
-    ) {
+    public sealed record Model(Account Account) {
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/Account/AccountBadgeBase.cs:34]: Record struct?

Comment (alexyakunin) [reply src/dotnet/UI.Blazor/Components/Account/AccountBadgeBase.cs:34]: It's less important here though - the other models I asked the same question about are used frequently.

File: src/dotnet/UI.Blazor/Components/Account/AccountCircle.razor [hunk 6]
```diff
@@ -2,10 +2,12 @@
 @inherits AccountBadgeBase
 @{
     var m = State.Value;
+    var badgeContent = ShowPresence ? @<AccountPresenceIndicator UserSid="@UserId" /> : (RenderFragment?)null;
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/Account/AccountCircle.razor:5]: Better split this into 3 lines.

File: src/dotnet/UI.Blazor/Components/Account/AccountPresenceIndicator.razor [hunk 7]
```diff
@@ -0,0 +1,43 @@
+﻿@namespace ActualChat.UI.Blazor.Components
+@using ActualChat.Users
+@inherits ComputedStateComponent<Presence>
+
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/Account/AccountPresenceIndicator.razor:4]: No empty line here.

File: src/dotnet/UI.Blazor/Components/Account/AccountPresenceIndicator.razor [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        return new ComputedState<Presence>.Options {
+            InitialValue = presence,
+            Category = GetStateCategory(),
+        };
+    }
+
+    protected override async Task<Presence> ComputeState(CancellationToken cancellationToken) {
+        if (UserId.IsNone)
+            return Presence.Unknown;
+        return await UserPresences.Get(UserId, cancellationToken);
```
Comment (alexyakunin) [inline src/dotnet/UI.Blazor/Components/Account/AccountPresenceIndicator.razor:41]: Either an empty line is missing or you need to convert it into a ternary "?:".

Comment (alexyakunin) [reply src/dotnet/UI.Blazor/Components/Account/AccountPresenceIndicator.razor:41]: Any `return` implies an empty line _after_, unless:
- It's the last line of a statement block (i.e. it follows with `}`)
- It's a `return` inside a sequence of `if (...) return` - in this case only the last return in such a sequence needs an empty line. Sorta fit here, but I'd rather use `?:`.

**Pull request merged**